Apply FluidProperties class to LiquidFlow

wenqing opened the pull request:
- Apply FluidProperties class to LiquidFlow
- Added a creator for LiquidFlowMaterialProperties

Note: This is the 2nd PR for the staggered scheme implementation.

TomFischer commented on ProcessLib/LiquidFlow/CreateLiquidFlowMaterialProperties.cpp:
```diff
@@ -0,0 +1,94 @@
+        storage_models.emplace_back(std::move(beta));
+    }
+
+    BaseLib::reorderVector(intrinsic_permeability_models, mat_ids);
```
I don't understand why the `intrinsic_permeability` (and `porosity_models` as well as `storage_models`) have to be reordered. Please, can you explain it with a small example what is done here?

wenqing replied:
Here is an example:
```
     <porous_medium>
          <porous_medium  id="0">
                <permeability>
                     ...
                 </permeability>
                 <porosity>
                    ...
                 </porosity>
                 <storage>
                   ...
                  </storage>
            </porous_medium>
          <porous_medium  id="1">
                <permeability>
                     ...
                 </permeability>
                 <porosity>
                    ...
                 </porosity>
                 <storage>
                   ...
                  </storage>
            </porous_medium>
            ...
        </porous_medium>
```
After parsing, we can not assure that parameters under ``id="N"`` are pushed to the ``N`` th element of the corresponding vector. Therefore a sorting is performed according to the IDs that are read from the tag attribution.

TomFischer replied:
Thank you for the explanation. Please add this as an explanation/comment  in the code.

TomFischer commented on ProcessLib/LiquidFlow/CreateLiquidFlowMaterialProperties.cpp:
```diff
@@ -0,0 +1,94 @@
+    {
+        //! \ogs_file_attr{prj__processes__process__LIQUID_FLOW__material_property__porous_medium__porous_medium__id}
+        auto const id = conf.getConfigAttributeOptional<int>("id");
+        mat_ids.push_back(*id);
```
The `id` seems to be an optional parameter. What is the purpose of this parameter and is it okay to insert `id` into the `mat_ids` when `id` is optional?

wenqing replied:
Removed the ``optional``. This parameter is the material ID, which is read from vtu mesh file.

TomFischer commented on ProcessLib/LiquidFlow/CreateLiquidFlowMaterialProperties.cpp:
```diff
@@ -0,0 +1,94 @@
+class LiquidFlowMaterialProperties;
+
+std::unique_ptr<LiquidFlowMaterialProperties>
+createLiquidFlowMaterialProperties(
```
The documentations for the parameters read in this function is missing. Probably the documentation will make the review a bit easier.

wenqing replied:
The parameters in this function are documented (see Documentation/prj/processes/process/LIQUID_FLOW). The parameters of material models are given in the creating functions of the models.

Actually, the body of this function is exactly moved from the constructor of class LiquidFlowMaterialProperties.

TomFischer replied:
Thanks for explanation.

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowMaterialProperties.cpp:
```diff
@@ -107,7 +61,8 @@ double LiquidFlowMaterialProperties::getViscosity(const double p,
     vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::p)] = p;
-    return _viscosity->getValue(vars);
+    return _fluid_properties->getValue(
+        MaterialLib::Fluid::FluidPropertyType::Vicosity, vars);
```
Spelling: `Vicosity` should be `Viscosity`.

wenqing replied:
Crrected.

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowMaterialProperties.h:
```diff
@@ -44,9 +53,23 @@ class LiquidFlowMaterialProperties
     typedef MaterialLib::Fluid::FluidProperty::ArrayType ArrayType;
 
     LiquidFlowMaterialProperties(
```
Is it correct to say: `LiquidFlowMaterialProperties` is the collection of material properties special to the `LiquidFlow` process? Maybe you can add some documentation?

wenqing replied:
Added a documentation about the class.

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowMaterialProperties.h:
```diff
@@ -79,10 +102,12 @@ class LiquidFlowMaterialProperties
 
     double getViscosity(const double p, const double T) const;
```
I thought that this class only collects porous media material properties and `LiquidFlowMaterialProperties` is responsible for viscosity and density. So probably my understanding is not correct. So my maby stupid question: Why is the viscosity in this class?

wenqing replied:
It is somehow a forward call to compute the viscosity. However, the arguments are p and T directly instead of an array of variable for this specific flow process.

Besides, I added a sentence of documentation about this class as
```
/** This class has a collection of material properties of liquid flow process
 *  and it provides the computation of the coefficients of the PDE of liquid
 *  flow process.
 */
```

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowProcess.h:
```diff
@@ -12,12 +12,17 @@
 #pragma once
 
+#include <memory>
+#include <memory>
```
You want to be really sure to `#include <memory>`? 😃

wenqing replied:
:smile: Be afraid of memory shortage.:smile:. Removed one.

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowProcess.h:
```diff
@@ -72,6 +77,14 @@ class LiquidFlowProcess final : public Process
+    std::vector<MaterialLib::Fluid::FluidProperties*> getFluidProperties() const
+    {
+        std::vector<MaterialLib::Fluid::FluidProperties*> material_properties =
+            {_material_properties->getFluidProperties()};
```
Verz minor: Why are the curly braces needed here?

TomFischer replied:
I could not find the place where `getFluidProperties()` is used? Why is it necessary to return a `std::vector<MaterialLib::Fluid::FluidProperties*>` and not only something like `MaterialLib::Fluid::FluidProperties const&`?

wenqing replied:
With that curly braces, I want to return a vector with one entry.

Initially, I want to have this function for all flow processes for coupling computation. Now I changed the mind, and made this function be unique only for LiquidFlow and return only a pointer to  class FluidProperties.

TomFischer replied:
Okay, but why is it necessary to return the pointer in a vector?

wenqing replied:
Originally it is returned in a vector in case to have a same return turn type with multiphase flow process, which can have more than one fluids. Now function is changed to
```
    MaterialLib::Fluid::FluidProperties* getFluidProperties() const
    {
        return _material_properties->getFluidProperties();
    }
```
for the LiquidFlow only.

TomFischer replied:
Thanks, now I got it.

TomFischer commented on ProcessLib/LiquidFlow/LiquidFlowProcess.h:
```diff
@@ -88,7 +101,7 @@ class LiquidFlowProcess final : public Process
     const int _gravitational_axis_id;
     const double _gravitational_acceleration;
-    LiquidFlowMaterialProperties _material_properties;
+    const std::unique_ptr<LiquidFlowMaterialProperties> _material_properties;
```
Why not a `const reference`? If it would be a `const&` maybe the implementation of `getFluidProperties()` above can return a `const& FluidProperties` and put it outside of the class into a `std::vector`?

wenqing replied:
As explained above,  getFluidProperties() is now changed to return a pointer. Therefore, I prefer to keep this declaration as a concrete instance. While the local assembler of LiquidFlow takes  a reference to this unique_ptr member.

wenqing commented:
@endJunction  Two commits according to your comments and my improvement.
- Commit  fcdeac7: Use full variable names in createLiquidFlowMaterialProperties.
- Commit 	1656b03: Changed the type of a member from unique_ptr reference to class reference
                               and removed a member, _current_material_id for thread safe.

endJunction commented on ProcessLib/LiquidFlow/LiquidFlowLocalAssembler.h:
```diff
@@ -61,21 +62,23 @@ class LiquidFlowLocalAssembler : public LiquidFlowLocalAssemblerInterface
+        unsigned const integration_order,
+        int const gravitational_axis_id,
+        double const gravitational_acceleration,
+        std::unique_ptr<LiquidFlowMaterialProperties> const&
```
Could be simply `LiquidFlowMaterialProperties const&`...

wenqing replied:
Changed. Accordingly, ``_material_properties`` in line 65 of LiquidFlowProcess.cpp is changed to ``*_material_properties``.

 See: https://github.com/ufz/ogs/pull/1631/commits/76dd96bcdf43d60b79d3f313f612d6a8c6f52f15#diff-97a96abf0d3aa134d63d918100ee67b7R65

endJunction commented on ProcessLib/LiquidFlow/LiquidFlowLocalAssembler.h:
```diff
@@ -61,21 +62,23 @@ class LiquidFlowLocalAssembler : public LiquidFlowLocalAssemblerInterface
           _gravitational_axis_id(gravitational_axis_id),
           _gravitational_acceleration(gravitational_acceleration),
-          _material_properties(material_propertries)
+          _material_properties(*(material_propertries.get()))
```
I just mentioned this somewhere else: `get` is not necessary because `material_properties` is a pointer, so simply dereferencing it is good.

wenqing replied:
Changed back to ``_material_properties(material_propertries)`` because the argument type ``std::unique_ptr<LiquidFlowMaterialProperties> const&`` has been changed to ``LiquidFlowMaterialProperties const&`` according to you last comment.

PS: I made these two changes by amending the last commit.

endJunction commented:
:fast_forward:
